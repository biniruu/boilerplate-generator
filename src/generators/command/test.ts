import getCertainConditions from '@utils/certainConditions'
import type { SelectOptions } from '_types'

const getTestCommands = (configOptions: SelectOptions) => {
  const { hasCssModule, hasJest, hasNext, hasNuxt, hasReact, hasStorybook, hasTypescript, hasVue, hasWebpack } =
    getCertainConditions(configOptions)

  const testDevDependencies: string[] = []

  if (hasJest) {
    /**
     * @testing-library/dom (DOM Testing Library)
     * {@link https://testing-library.com/docs/dom-testing-library/intro/}
     *
     * @testing-library/jest-dom (jest-dom)
     * {@link https://github.com/testing-library/jest-dom#readme}
     *
     * @testing-library/user-event (user-event)
     * {@link https://github.com/testing-library/user-event#readme}
     *
     * jest (Jest)
     * {@link https://jestjs.io}
     *
     * jest-environment-jsdom
     * {@link https://github.com/jestjs/jest#readme}
     *
     * Jest Extended
     * {{@link https://jest-extended.jestcommunity.dev/docs}}
     *
     * jest-watch-typeahead
     * {@link https://github.com/jest-community/jest-watch-typeahead#jest-watch-typeahead}
     *
     * jsdom
     * {@link https://github.com/jsdom/jsdom#readme}
     *
     * msw (Mock Service Worker)
     * {@link https://mswjs.io}
     *
     * ts-jest
     * {@link https://kulshekhar.github.io/ts-jest/}
     * ts-jest is necessary when youâ€™re working in a development environment, since it provides types of jest configuration
     *
     *
     * If you have added 'jest', you don't need to add these packages: jest-resolve and jest-resolve-dependencies.
     * I recommend adding @types/jest even if it is not a TypeScript project because of its IntelliSense.
     */
    testDevDependencies.push(
      '@testing-library/dom',
      '@testing-library/jest-dom',
      '@testing-library/user-event',
      '@types/jest',
      'jest',
      'jest-environment-jsdom',
      'jest-extended',
      'jest-html-reporters',
      'jest-watch-typeahead',
      'jsdom',
      'msw',
      'ts-jest',
    )
    /**
     * identity-obj-proxy
     * {@link https://github.com/keyz/identity-obj-proxy#readme}
     */
    if (hasCssModule) {
      if (hasTypescript) {
        testDevDependencies.push('@types/identity-obj-proxy')
      }
      testDevDependencies.push('identity-obj-proxy')
    }
    /**
     * @testing-library/react (React Testing Library)
     * {@link https://testing-library.com/docs/react-testing-library/intro/})
     *
     * @testing-library/react-hooks (react-hooks-testing-library)
     * {@link https://github.com/testing-library/react-hooks-testing-library#readme}
     *
     * react-test-renderer (Test Renderer)
     * {@link https://legacy.reactjs.org/docs/test-renderer.html}
     */
    if (hasReact || hasNext) {
      if (hasStorybook) {
        testDevDependencies.push('@storybook/react', '@storybook/testing-library')
      }
      testDevDependencies.push(
        '@testing-library/react',
        '@testing-library/react-hooks',
        '@types/react-test-renderer',
        'react-test-renderer',
      )
    }
    /**
     * @testing-library/vue (Vue Testing Library)
     * {@link https://testing-library.com/docs/vue-testing-library/intro/}
     *
     * @vue/vue3-jest
     * {@link https://github.com/vuejs/vue-jest#readme}
     */
    if (hasNuxt || hasVue) {
      testDevDependencies.push('@testing-library/vue', '@vue/vue3-jest')
    }
    // TODO: Edit this jsdoc
    /**
     * @storybook/addon-a11y (Accessibility testing)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-actions (Action logger)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-essentials (Essential addons)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-interactions (Interact with components)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-links (Link to other stories)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/builder-webpack5 (Webpack 5)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/manager-webpack5 (Webpack 5 Storybook manager)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     */
    if (hasStorybook && !hasReact && !hasNext) {
      if (hasWebpack) {
        testDevDependencies.push('@storybook/builder-webpack5', '@storybook/manager-webpack5')
      }
      testDevDependencies.push(
        '@storybook/addon-a11y',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-interactions',
        '@storybook/addon-links',
      )
    }
  }

  return {
    testDevDependencies,
  }
}

export default getTestCommands
