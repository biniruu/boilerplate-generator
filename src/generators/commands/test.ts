import type { SelectOptions } from '_types'

const getTestCommands = (configOptions: SelectOptions) => {
  const hasJest = configOptions.jest
  const hasNext = configOptions.next
  const hasNuxt = configOptions.nuxt
  const hasPostcss = configOptions.postcss
  const hasReact = configOptions.react
  const hasTailwind = configOptions.tailwind
  const hasTypescript = configOptions.typescript
  const hasVue = configOptions.vue
  const hasCssModule = hasPostcss || hasTailwind

  const testDevDependencies: string[] = []

  if (hasJest) {
    if (hasCssModule) {
      if (hasTypescript) {
        testDevDependencies.push('@types/identity-obj-proxy')
      }
      testDevDependencies.push('identity-obj-proxy')
    }
    if (hasReact || hasNext) {
      testDevDependencies.push(
        '@testing-library/react',
        '@testing-library/react-hooks',
        '@types/react-test-renderer',
        'react-test-renderer',
      )
    }
    if (hasTypescript) {
      testDevDependencies.push('ts-jest')
    }
    if (hasNuxt || hasVue) {
      testDevDependencies.push('@testing-library/vue', '@vue/vue3-jest')
    }
    // If you have added 'jest', you don't need to add these packages: jest-resolve and jest-resolve-dependencies.
    // I recommend adding @types/jest even if it is not a TypeScript project because of its IntelliSense.
    testDevDependencies.push(
      '@testing-library/dom',
      '@testing-library/jest-dom',
      '@testing-library/user-event',
      '@types/jest',
      'jest',
      'jest-environment-jsdom',
      'jest-watch-typeahead',
      'jsdom',
      'msw',
    )
  }

  return {
    testDevDependencies,
  }
}

export default getTestCommands
