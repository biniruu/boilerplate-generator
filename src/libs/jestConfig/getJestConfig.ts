import config from '@data/jest/jest.json'
import mergeConfigs from '@utils/mergeConfigs'

type Syntax = 'typescript' | 'javascript'

const data: { [key: string]: unknown } = {
  typescript: await import('@data/jest/jest-ts.json'),
}

const transform = {
  typescript: {
    transform: {
      '^.+\\.(ts|tsx)$': [
        'ts-jest',
        {
          tsconfig: './tsconfig.test.json',
        },
      ],
    },
  },
  javascript: {
    transform: {
      '^.+\\.jsx?$': 'babel-jest',
    },
  },
}

const addTransform = (config: { [key: string]: unknown }, selectedVal: Syntax) => {
  return { ...config, ...transform[selectedVal] }
}

const getJestConfig = (selectedVal: Syntax) => {
  if (selectedVal === 'typescript') {
    const newConfig = { ...config }
    mergeConfigs(newConfig, 'preset', data[selectedVal] as { [key: string]: unknown })

    return addTransform(newConfig, selectedVal)
  }

  return addTransform(config, selectedVal)
}

export default getJestConfig
