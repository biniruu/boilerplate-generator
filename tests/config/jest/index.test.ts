import getJestConfig from '@generators/config/jest/jest'

import { elemCode, options, setTextContent } from '../setTextContent.test'

describe('For jest.config.ts', () => {
  it('should return the configuration that has no selected options', () => {
    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })

  it('should return the configuration with TypeScript', () => {
    options.typescript = true

    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })

  it('should return the configuration with Lodash', () => {
    options.lodash = true

    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })

  it('should return the configuration with Next.js', () => {
    options.next = true

    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })

  it('should return the configuration with Next.js and TypeScript', () => {
    options.next = true
    options.typescript = true

    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })

  it('should return the configuration with Vue.js', () => {
    options.next = true

    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })

  it('should return the configuration with Jest HTML Reporters', () => {
    options.jestHtmlReporters = true

    setTextContent(getJestConfig)

    expect(elemCode?.textContent).toMatchSnapshot()
  })
})
