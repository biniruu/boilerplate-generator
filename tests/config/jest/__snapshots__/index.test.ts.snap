// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`For jest.config.ts should return the configuration that has no selected options 1`] = `
"/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
    
module.exports = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy'
  },
  reporters: [
    'default'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.js',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/'
  ],
  transform: {
    '^.+\\\\.(js|jsx)$': 'babel-jest'
  },
  verbose: true
}"
`;

exports[`For jest.config.ts should return the configuration with Jest HTML Reporters 1`] = `
"/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
    
module.exports = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy'
  },
  reporters: [
    'default',
    'jest-html-reporters'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.js',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/'
  ],
  transform: {
    '^.+\\\\.(js|jsx)$': 'babel-jest'
  },
  verbose: true
}"
`;

exports[`For jest.config.ts should return the configuration with Lodash 1`] = `
"/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
    
module.exports = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy',
    'lodash-es': 'lodash'
  },
  reporters: [
    'default'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.js',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/'
  ],
  transform: {
    '^.+\\\\.(js|jsx)$': 'babel-jest'
  },
  verbose: true
}"
`;

exports[`For jest.config.ts should return the configuration with Next.js 1`] = `
"/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
    
module.exports = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy'
  },
  reporters: [
    'default'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.js',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/',
    './.next/'
  ],
  transform: {
    '^.+\\\\.(js|jsx)$': 'babel-jest'
  },
  verbose: true
}"
`;

exports[`For jest.config.ts should return the configuration with Next.js and TypeScript 1`] = `
"import type { JestConfigWithTsJest } from 'ts-jest'

import 'jest-extended'
    
const jestConfig: JestConfigWithTsJest = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy'
  },
  preset: 'ts-jest',
  reporters: [
    'default'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.ts',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/',
    './.next/'
  ],
  transform: {
    '^.+\\\\.(ts|tsx)$': [
      'ts-jest',
      {
        tsconfig: './tsconfig.test.json'
      }
    ]
  },
  verbose: true
}

export default jestConfig"
`;

exports[`For jest.config.ts should return the configuration with TypeScript 1`] = `
"import type { JestConfigWithTsJest } from 'ts-jest'

import 'jest-extended'
    
const jestConfig: JestConfigWithTsJest = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy'
  },
  preset: 'ts-jest',
  reporters: [
    'default'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.ts',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/'
  ],
  transform: {
    '^.+\\\\.(ts|tsx)$': [
      'ts-jest',
      {
        tsconfig: './tsconfig.test.json'
      }
    ]
  },
  verbose: true
}

export default jestConfig"
`;

exports[`For jest.config.ts should return the configuration with Vue.js 1`] = `
"/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
    
module.exports = {
  clearMocks: true,
  collectCoverage: true,
  coverageDirectory: './coverage',
  moduleDirectories: [
    './node_modules',
    './src'
  ],
  moduleNameMapper: {
    '@app/(.*)': './src/app/$1',
    '\\\\.(css|scss)$': 'identity-obj-proxy'
  },
  reporters: [
    'default'
  ],
  rootDir: '.',
  roots: [
    'tests'
  ],
  setupFilesAfterEnv: [
    './jest.setup.js',
    'jest-extended/all'
  ],
  testEnvironment: 'jest-environment-jsdom',
  testMatch: [
    '**/tests/**/*.[jt]s?(x)',
    '**/__tests__/**/*.[jt]s?(x)',
    '**/?(*.)+(spec|test).[tj]s?(x)'
  ],
  testPathIgnorePatterns: [
    './node_modules/',
    './.next/'
  ],
  transform: {
    '^.+\\\\.(js|jsx)$': 'babel-jest'
  },
  verbose: true
}"
`;
